File: backend/nodejs-typescript/src/modules/contact/contact.service.ts
import { ContactDto } from "./contact.schema";
import { NotFoundError, ValidationError } from "@/utils/response-error.util";
import Contact from "@/models/contact";

type AddressType = {
  id: number;
  street: string;
  city: string;
  district: string;
  subDistrict: string;
  postalCode: string;
};

type ContactType = {
  id: number;
  contactId: number;
  fullName: string;
  nickName: string | null;
  phoneNumber: string;
  email: string | null;
  addresses: AddressType[];
};

export default class ContactService {
  static async getContacts(user: Express.User) {
    const contacts = await Contact.query()
      .select(
        "id",
        "full_name as fullName",
        "nick_name as nickName",
        "phone_number as phoneNumber",
        "email",
      )
      .where("user_id", user.id)
      .withGraphFetched("addresses")
      .modifyGraph("addresses", (builder) => {
        builder.select(
          "id",
          "contact_id as contactId",
          "street",
          "city",
          "district",
          "sub_district as subDistrict",
          "postal_code as postalCode",
        );
      })
      .castTo<ContactType[]>();

    return contacts;
  }

  static async getContact(user: Express.User, contactId: number) {
    const contact = await Contact.query()
      .select(
        "id",
        "full_name as fullName",
        "nick_name as nickName",
        "phone_number as phoneNumber",
        "email",
      )
      .where("user_id", user.id)
      .andWhere("id", contactId)
      .withGraphFetched("addresses")
      .modifyGraph("addresses", (builder) => {
        builder.select(
          "id",
          "contact_id as contactId",
          "street",
          "city",
          "district",
          "sub_district as subDistrict",
          "postal_code as postalCode",
        );
      })
      .first()
      .castTo<ContactType>();

    if (!contact) {
      throw new NotFoundError("Contact not found");
    }

    return contact;
  }

  static async addContact(user: Express.User, contactDto: ContactDto) {
    await Contact.query().insertGraph({
      user_id: user.id,
      full_name: contactDto.fullName,
      nick_name: contactDto.nickName,
      phone_number: contactDto.phoneNumber,
      email: contactDto.email,
      addresses: contactDto.addresses.map((address) => ({
        street: address.street,
        city: address.city,
        district: address.district,
        sub_district: address.subDistrict,
        postal_code: address.postalCode,
      })),
    });
  }

  static async updateContact(
    user: Express.User,
    contactId: number,
    contactDto: ContactDto,
  ) {
    const contact = await Contact.query()
      .where("user_id", user.id)
      .andWhere("id", contactId)
      .first();

    if (!contact) {
      throw new NotFoundError("Contact not found");
    }

    await Contact.query().upsertGraph({
      id: contact.id,
      user_id: user.id,
      full_name: contactDto.fullName,
      nick_name: contactDto.nickName,
      phone_number: contactDto.phoneNumber,
      email: contactDto.email,
      addresses: contactDto.addresses.map((address) => ({
        street: address.street,
        city: address.city,
        district: address.district,
        sub_district: address.subDistrict,
        postal_code: address.postalCode,
      })),
    });
  }

  static async deleteContact(user: Express.User, contactId: number) {
    const contact = await Contact.query()
      .where("user_id", user.id)
      .andWhere("id", contactId)
      .first();

    if (!contact) {
      throw new NotFoundError("Contact not found");
    }

    await Contact.transaction(async (trx) => {
      await Contact.relatedQuery("addresses", trx).for(contact.id).delete();
      await Contact.relatedQuery("groups", trx).for(contact.id).unrelate();
      await Contact.query(trx).where("id", contact.id).delete();
    });
  }
}

File: backend/go/internal/services/contact_service.go
package services

import (
	"github.com/fathurmdr/backend/go/internal/db"
	"github.com/fathurmdr/backend/go/internal/dto"
	"github.com/fathurmdr/backend/go/internal/models"
	"github.com/fathurmdr/backend/go/internal/utils"
	"gorm.io/gorm"
)


type ContactService struct{}

func (contactService *ContactService) GetContacts(user models.User) (*[]dto.Contact, error) {
	var contacts []models.Contact

	err := db.DB.Model(&models.Contact{}).
			Preload("Addresses", func(db *gorm.DB) *gorm.DB {
					return db.Select("ID", "ContactID", "Street", "City", "District", "SubDistrict", "PostalCode")
			}).
			Select("ID", "FullName", "NickName", "PhoneNumber", "Email").
			Where("user_id = ?", user.ID).
			Find(&contacts).Error

	if err != nil {
			return nil,  utils.NewApplicationError("", nil)
	}

	contactsResponse := make([]dto.Contact,0)

	for _, contact := range contacts {
		addresses := make([]dto.Address,0)
		for _, address := range contact.Addresses {
			addresses = append(addresses, dto.Address{
				ID:          address.ID,
				ContactID:   address.ContactID,
				Street:      address.Street,
				City:        address.City,
				District:    address.District,
				SubDistrict: address.SubDistrict,
				PostalCode:  address.PostalCode,
			})
		}
		contactsResponse = append(contactsResponse, dto.Contact{
			ID:          contact.ID,
			FullName:    contact.FullName,
			NickName:    contact.NickName,
			PhoneNumber: contact.PhoneNumber,
			Email:       contact.Email,
			Addresses:   addresses,
		})
	}

	return &contactsResponse, nil
}

func (contactService *ContactService) GetContact(user models.User, id uint) (*dto.Contact, error) {

	var contact models.Contact

	err := db.DB.
				Preload("Addresses", func(db *gorm.DB) *gorm.DB {
						return db.Select("ID", "ContactID", "Street", "City", "District", "SubDistrict", "PostalCode")
				}).
				Select("ID", "FullName", "NickName", "PhoneNumber", "Email").
				Where("user_id = ?", user.ID).First(&contact).Error
	if err != nil {
		return nil, utils.NewNotFoundError("Contact not found", nil)
	}


	addresses := make([]dto.Address,0)
	for _, address := range contact.Addresses {
		addresses = append(addresses, dto.Address{
			ID:          address.ID,
			ContactID:   address.ContactID,
			Street:      address.Street,
			City:        address.City,
			District:    address.District,
			SubDistrict: address.SubDistrict,
			PostalCode:  address.PostalCode,
		})
	}
	
	contactResponse := dto.Contact{
		ID:          contact.ID,
		FullName:    contact.FullName,
		NickName:    contact.NickName,
		PhoneNumber: contact.PhoneNumber,
		Email:       contact.Email,
		Addresses:   addresses,
	}

	return &contactResponse, nil
}

func (contactService *ContactService) AddContact(user models.User, contactRequest *dto.Contact) (error) {
	contact := models.Contact{
		UserID:      user.ID,
		FullName:    contactRequest.FullName,
		NickName:    contactRequest.NickName,
		PhoneNumber: contactRequest.PhoneNumber,
		Email:       contactRequest.Email,
	}

	for _, address := range contactRequest.Addresses {
		contact.Addresses = append(contact.Addresses, models.Address{
			Street: address.Street,
			City: address.City,
			District: address.District,
			SubDistrict: address.SubDistrict,
			PostalCode: address.PostalCode,
		})
	}


	err := db.DB.Create(&contact).Error
	if err != nil {
		return utils.NewApplicationError("", nil)
	}

	return nil
}

func (contactService *ContactService) UpdateContact(user models.User, contactID uint, contactRequest *dto.Contact) (error) {
	var contact models.Contact
	err := db.DB.Where("id = ? AND user_id = ?", contactID, user.ID).First(&contact).Error
	if err != nil {
		return utils.NewNotFoundError("Contact not found", nil)
	}

	contact.FullName = contactRequest.FullName
	contact.NickName = contactRequest.NickName
	contact.PhoneNumber = contactRequest.PhoneNumber
	contact.Email = contactRequest.Email

	err = db.DB.Where("contact_id = ?", contact.ID).Delete(&models.Address{}).Error
	if err != nil {
		return utils.NewApplicationError("", nil)
	}

	for _, address := range contactRequest.Addresses {
		contact.Addresses = append(contact.Addresses, models.Address{
			Street:      address.Street,
			City:        address.City,
			District:    address.District,
			SubDistrict: address.SubDistrict,
			PostalCode:  address.PostalCode,
		})
	}

	err = db.DB.Save(&contact).Error
	if err != nil {
		return utils.NewApplicationError("", nil)
	}

	return nil
}

func (contactService *ContactService) DeleteContact(user models.User, contactID uint) (error) {

	var contact models.Contact
	err := db.DB.Where("id = ? AND user_id = ?", contactID, user.ID).First(&contact).Error
	if err != nil {
		return utils.NewNotFoundError("Contact not found", nil)
	}

	err = db.DB.Transaction(func(tx *gorm.DB) error {
		var address models.Address
		err = tx.Where("contact_id = ?", contact.ID).Delete(&address).Error
		if err != nil {
			return err
		}

		err = tx.Model(&contact).Association("Groups").Clear()
		if err != nil {
			return err
		}

		err = tx.Delete(&contact).Error
		if err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		return utils.NewApplicationError("Failed to delete contact", nil)
	}

	return nil
}
